swupd/hash.go

```diff
@@ -34,16 +35,26 @@ var AllZeroHash = "0000000000000000000000000000000000000000000000000000000000000
 // Hashes is a global map of indices to hashes
 var Hashes = []*string{&AllZeroHash}
 var invHash = map[string]Hashval{AllZeroHash: 0}
+var rwMutex sync.RWMutex
 
 // internHash adds only new hashes to the Hashes slice and returns the index at
 // which they are located
 func internHash(hash string) Hashval {
+	rwMutex.RLock()
 	if key, ok := invHash[hash]; ok {
+		rwMutex.RUnlock()
```

Comment:
This block can be:
```go
key, ok := invHash[hash]
rwMutex.Runlock()
if ok {
    return key
}
rwMutex.Lock()
...
```
So you don't have the multiple calls to `RUnlock` and you perform the unlock before checking your conditional.